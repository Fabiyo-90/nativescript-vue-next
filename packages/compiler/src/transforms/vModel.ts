import {
  DirectiveTransform,
  NodeTypes,
  transformModel as baseTransform,
} from '@vue/compiler-core'
import { V_MODEL } from '@nativescript-vue/compiler'

export const transformModel: DirectiveTransform = (dir, node, context) => {
  const baseResult = baseTransform(dir, node, context)
  baseResult.needRuntime = context.helper(V_MODEL)
  baseResult.props = baseResult.props.filter((p) => {
    if (
      p.key.type === NodeTypes.SIMPLE_EXPRESSION &&
      p.key.content === 'modelValue'
    ) {
      return false
    }
    return true
  })
  return baseResult
}
