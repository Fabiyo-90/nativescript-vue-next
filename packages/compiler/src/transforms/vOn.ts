import {
  transformOn as baseTransform,
  DirectiveTransform,
  createObjectProperty,
  createSimpleExpression,
  createCompoundExpression,
  isStaticExp
} from '@vue/compiler-core'
import { capitalize, makeMap } from '@vue/shared'

// todo: simplify - we only support `once` - no need to use makeMap
const isEventOptionModifier = /*#__PURE__*/ makeMap(`once,capture`)

// todo: remove - no need to loop through modifiers as only `once` is supported
const generateOptionModifiers = (modifiers: string[]) => {
  const eventOptionModifiers = []

  for (let i = 0; i < modifiers.length; i++) {
    const modifier = modifiers[i]

    if (isEventOptionModifier(modifier)) {
      // eventOptionModifiers: modifiers for addEventListener() options, e.g. .once
      eventOptionModifiers.push(modifier)
    }
  }

  return eventOptionModifiers
}

export const transformOn: DirectiveTransform = (dir, node, context) => {
  return baseTransform(dir, node, context, baseResult => {
    const { modifiers } = dir
    if (!modifiers.length) return baseResult

    let { key, value: handlerExp } = baseResult.props[0]
    const eventOptionModifiers = generateOptionModifiers(modifiers)

    if (eventOptionModifiers.length) {
      const modifierPostfix = eventOptionModifiers.map(capitalize).join('')
      key = isStaticExp(key)
        ? createSimpleExpression(`${key.content}${modifierPostfix}`, true)
        : createCompoundExpression([`(`, key, `) + "${modifierPostfix}"`])
    }

    return {
      props: [createObjectProperty(key, handlerExp)]
    }
  })
}
