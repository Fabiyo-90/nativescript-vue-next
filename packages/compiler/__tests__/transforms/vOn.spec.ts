import {
  baseParse as parse,
  transform,
  CompilerOptions,
  ElementNode,
  ObjectExpression,
  VNodeCall,
  transformElement,
  transformExpression
} from '@vue/compiler-core'
import { transformOn } from '../../src/transforms/vOn'

function parseWithVOn(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement],
    directiveTransforms: {
      on: transformOn
    },
    ...options
  })
  return {
    root: ast,
    props: (((ast.children[0] as ElementNode).codegenNode as VNodeCall)
      .props as ObjectExpression).properties
  }
}

describe('compiler: transform v-on', () => {
  it('should work w/ prefixIdentifiers: true', () => {
    const {
      props: [prop]
    } = parseWithVOn(`<Label @tap.once="test"/>`, {
      prefixIdentifiers: true
    })
    expect(prop).toMatchSnapshot()
  })
})
