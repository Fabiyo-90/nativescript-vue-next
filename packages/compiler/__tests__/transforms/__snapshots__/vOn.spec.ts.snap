// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: transform v-on should work w/ prefixIdentifiers: true 1`] = `
Object {
  "key": Object {
    "content": "onTap",
    "isConstant": false,
    "isStatic": true,
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": "tap",
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 4,
  },
  "loc": Object {
    "end": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "source": "",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": 16,
  "value": Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "source": "",
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "properties": Array [
      Object {
        "key": Object {
          "content": "handler",
          "isConstant": false,
          "isStatic": true,
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
            "source": "",
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": 4,
        },
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
          "source": "",
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "type": 16,
        "value": Object {
          "content": "_ctx.test",
          "isConstant": false,
          "isStatic": false,
          "loc": Object {
            "end": Object {
              "column": 23,
              "line": 1,
              "offset": 22,
            },
            "source": "test",
            "start": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
          },
          "type": 4,
        },
      },
      Object {
        "key": Object {
          "content": "options",
          "isConstant": false,
          "isStatic": true,
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
            "source": "",
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": 4,
        },
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
          "source": "",
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "type": 16,
        "value": Object {
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
            "source": "",
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "properties": Array [
            Object {
              "key": Object {
                "content": "once",
                "isConstant": false,
                "isStatic": true,
                "loc": Object {
                  "end": Object {
                    "column": 1,
                    "line": 1,
                    "offset": 0,
                  },
                  "source": "",
                  "start": Object {
                    "column": 1,
                    "line": 1,
                    "offset": 0,
                  },
                },
                "type": 4,
              },
              "loc": Object {
                "end": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
                "source": "",
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": 16,
              "value": Object {
                "content": "true",
                "isConstant": false,
                "isStatic": false,
                "loc": Object {
                  "end": Object {
                    "column": 1,
                    "line": 1,
                    "offset": 0,
                  },
                  "source": "",
                  "start": Object {
                    "column": 1,
                    "line": 1,
                    "offset": 0,
                  },
                },
                "type": 4,
              },
            },
          ],
          "type": 15,
        },
      },
    ],
    "type": 15,
  },
}
`;
